vmm/src/landlock.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright © 2024 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use landlock::{
+    path_beneath_rules, Access, AccessFs, BitFlags, PathFdError, Ruleset, RulesetAttr,
+    RulesetCreated, RulesetCreatedAttr, RulesetError, ABI,
+};
+use std::io::Error as IoError;
+use std::path::PathBuf;
+use thiserror::Error;
+
+use crate::vm_config::LandlockConfig;
+
+pub const READ: u8 = 1 << 0;
+pub const WRITE: u8 = 1 << 1;
+pub const EXECUTE: u8 = 1 << 2;
+
+#[derive(Debug, Error)]
+pub enum LandlockError {
+    /// Failed to create Ruleset
+    #[error("Error creating ruleset: {0}")]
+    CreateRuleSet(#[source] RulesetError),
+
+    /// Failed to add rule
+    #[error("Error adding rule to ruleset: {0}")]
+    AddRule(#[source] RulesetError),
+
+    /// Failed to restrict self
+    #[error("Error restricting self: {0}")]
+    RestrictSelf(#[source] RulesetError),
+
+    /// Ruleset is empty
+    #[error("Ruleset was previously consumed, cannot add rules to it")]
+    RulesetConsumed,
+
+    /// Error opening Path
+    #[error("Error opening Path: {0}")]
+    OpenPath(#[source] PathFdError),
+
+    /// Error opening Device Path
+    #[error("Error opening Device Path: {0}")]
+    OpenDevicePath(#[source] IoError),
+}
+
+pub struct Landlock {
+    ruleset: Option<RulesetCreated>,
+    abi: ABI,
+}
+
+impl Landlock {
+    pub fn new() -> Result<Landlock, LandlockError> {
+        let abi = ABI::V3;
+
+        let file_access = AccessFs::from_all(abi);
+        let def_ruleset = Ruleset::default()
+            .handle_access(file_access)
+            .map_err(LandlockError::CreateRuleSet)?;
+
+        let ruleset = def_ruleset.create().map_err(LandlockError::CreateRuleSet)?;
+
+        Ok(Landlock {
+            ruleset: Some(ruleset),
+            abi,
+        })
+    }
+
+    pub fn add_rule(
+        &mut self,
+        path: PathBuf,
+        access: BitFlags<AccessFs>,
+    ) -> Result<(), LandlockError> {
+        if self.ruleset.is_none() {
+            return Err(LandlockError::RulesetConsumed);
+        }
+        // path_beneath_rules in landlock crate handles file and directory access rules.
+        // Incoming path/s are passed to path_beneath_rules, so that we don't
+        // have to worry about the type of the path.
+        let paths = vec![path.clone()];
+        let path_beneath_rules = path_beneath_rules(paths, access);
+        for rule in path_beneath_rules {
+            self.ruleset
+                .as_mut()
```

Comment:
Do we really need an explicit call to `as_mut()`?

Reply:
As ruleset is of type `Option<RulesetCreated>`, I need to use `as_mut()`. Without it I get the following error:
```
error[E0507]: cannot move out of `self.ruleset` which is behind a mutable reference
   --> vmm/src/landlock.rs:77:13
    |
77  |             self.ruleset
    |             ^^^^^^^^^^^^
    |             |
    |             help: consider calling `.as_ref()` or `.as_mut()` to borrow the type's contents
    |             move occurs because `self.ruleset` has type `std::option::Option<RulesetCreated>`, which does not implement the `Copy` trait
78  |                 .unwrap()
    |                  -------- `self.ruleset` moved due to this method call
    |
note: `std::option::Option::<T>::unwrap` takes ownership of the receiver `self`, which moves `self.ruleset`
```

Reply:
As shown in previous comment, with the use of add_rules, don't need `as_mut()`.

Reply:
done